components/org.wso2.carbon.identity.cors.valve/src/main/java/org/wso2/carbon/identity/cors/valve/CORSValve.java

```diff
@@ -0,0 +1,126 @@
+/**
+ * Valve class for CORS.
+ */
+public class CORSValve extends ValveBase {
+
+    private static final Log log = LogFactory.getLog(CORSValve.class);
+
+    private final CORSRequestHandler corsRequestHandler;
+
+    /**
+     * Default constructor for the CORSValve.
+     */
+    public CORSValve() {
+
+        corsRequestHandler = new CORSRequestHandler();
+    }
+
+    @Override
+    public void invoke(Request request, Response response) throws IOException, ServletException {
+
+        // Determine the type of the request.
+        CORSRequestType corsRequestType = CORSUtils.getRequestType(request);
+
+        try {
+            // Tag if configured.
+            CORSManager corsManager = CORSValveServiceHolder.getInstance().getCorsManager();
+            String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            CORSConfiguration corsConfiguration = corsManager.getCORSConfiguration(tenantDomain);
+            if (corsConfiguration.isTagRequests()) {
+                RequestTagger.tag(request, corsRequestType);
+            }
+
+            CORSConfiguration config = CORSValveServiceHolder.getInstance().getCorsManager()
+                    .getCORSConfiguration(tenantDomain);
+
+            if (corsRequestType == CORSRequestType.ACTUAL) {
+                // ACTUAL request.
+                corsRequestHandler.handleActualRequest(request, response);
+
+                getNext().invoke(request, response);
+            } else if (corsRequestType == CORSRequestType.PREFLIGHT) {
+                // PREFLIGHT request - Handle but don't pass further down the chain.
+                corsRequestHandler.handlePreflightRequest(request, response);
+            } else if (config.isAllowGenericHttpRequests()) {
+                // OTHER request - Not a CORS request, but allow it through.
+                getNext().invoke(request, response);
+            } else {
+                // Generic HTTP requests denied.
+                printMessage(new CORSException(ErrorMessages.ERROR_CODE_GENERIC_HTTP_NOT_ALLOWED), response);
+            }
+        } catch (CORSException e) {
+            printMessage(e, response);
+        } catch (CORSManagementServiceException e) {
+            if (log.isDebugEnabled()) {
+                log.debug(e);
+            }
+        }
+    }
+
+    /**
+     * Produces a simple HTTP text/plain response for the specified CORS
+     * exception.
+     * Note: The CORS filter avoids falling back to the default web container error page (typically a
+     * richly-formatted HTML page) to make it easier for XHR debugger tools to identify the cause of failed requests.
+     *
+     * @param corsException The CORS exception. Must not be {@code null}.
+     * @param response      The HTTP servlet response. Must not be {@code null}.
+     * @throws IOException On a I/O exception.
+     */
+    private void printMessage(final CORSException corsException, final HttpServletResponse response)
```

Comment:
inline 114 and 115

Reply:
If I inline this, line length would be too long.